File: AVP Lib/Drivers/RSTi/PropertyObject.vb
Imports System.Reflection

Public Class PropertyObject
    Public Enum DataType
        Undefined = 0
        IOLib_Base_EquipmentStatus
        System_Single
        System_Double
        System_Boolean
        IOLib_Base_MatchMode
    End Enum

    Private m_strPropertyName As String = String.Empty
    Private m_obj As Object = Nothing
    Private m_prop As PropertyInfo = Nothing
    Private m_bIsOk As Boolean = False
    Private m_eDataType As DataType = DataType.Undefined
    Private m_RawValue As String = String.Empty

    Public ReadOnly Property PropDataType() As DataType
        Get
            Return m_eDataType
        End Get
    End Property

    Public ReadOnly Property IsOK() As Boolean
        Get
            Return m_bIsOk
        End Get
    End Property

    Public Property RawValue() As String
        Get
            Return m_RawValue
        End Get
        Set(ByVal value As String)
            m_RawValue = value
        End Set
    End Property

    Public Sub New()
    End Sub

    Public Function Init(ByVal obj As DataManagerment.Equipment, ByVal name As String, ByVal ParamArray arrParams As Object()) As Boolean
        Dim prop As PropertyInfo = obj.GetType().GetProperty(name)


        If prop Is Nothing Then
            m_bIsOk = False
            Return False
        End If

        If Not prop.CanWrite Then
            m_bIsOk = False
            Return False
        End If

        m_strPropertyName = name
        m_prop = prop
        m_obj = obj
        m_bIsOk = True
        m_eDataType = GetDataType(m_prop.PropertyType.FullName)

        Return True
    End Function

    Public Sub SetValue(ByVal objValue As Object)
        Try

            If m_bIsOk Then
                'm_prop.SetValue(m_obj, objValue, Nothing)
            End If

        Catch ex As Exception
        End Try
    End Sub

    Public Function GetValue() As Object
        Try
            If m_prop IsNot Nothing Then
                Return m_prop.GetValue(m_obj, Nothing)
            End If

        Catch ex As System.Exception
        End Try

        Return Nothing
    End Function

    Private Function GetDataType(ByVal strDataType As String) As DataType
        Dim eDataType As DataType = DataType.Undefined

        If strDataType = "IOLib.Base.EquipmentStatus" Then
            eDataType = DataType.IOLib_Base_EquipmentStatus
        ElseIf strDataType = "System.Single" Then
            eDataType = DataType.System_Single
        ElseIf strDataType = "System.Double" Then
            eDataType = DataType.System_Double
        ElseIf strDataType = "System.Boolean" Then
            eDataType = DataType.System_Boolean
        ElseIf strDataType = "IOLib.Base.MatchMode" Then
            eDataType = DataType.IOLib_Base_MatchMode
        Else
            eDataType = DataType.Undefined
        End If

        Return eDataType
    End Function
End Class

File: DeviceNetApp/DeviceNetApp/Lib/PropertyObject.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Reflection;

namespace DeviceNetApp.Lib
{
    public class PropertyObject
    {
        public enum DataType
        {
            Undefined = 0,
            IOLib_Base_EquipmentStatus,
            System_Single,
            System_Boolean,
        }

        String m_strPropertyName = String.Empty;
        Object m_obj = null;
        PropertyInfo m_prop = null;
        Boolean m_bIsOk = false;
        DataType m_eDataType = DataType.Undefined;              

        public DataType PropDataType
        {
            get { return m_eDataType; }
        }

        public PropertyObject()
        {

        }

        public Boolean Init(Object obj, String name)
        {
            PropertyInfo prop = obj.GetType().GetProperty(name);
            if (prop == null)
            {
                m_bIsOk = false;
                return false;
            }
            if (!prop.CanWrite)
            {
                m_bIsOk = false;
                return false;
            }
            m_strPropertyName = name;
            m_prop = prop;
            m_obj = obj;
            m_bIsOk = true;
            m_eDataType = GetDataType(m_prop.PropertyType.FullName);

            return true;
        }

        public void SetValue(Object objValue)
        {
            try
            {
                if (m_bIsOk)
                {
                    m_prop.SetValue(m_obj, objValue, null);
                }
            }
            catch (Exception ex)
            {
                System.Diagnostics.Debug.Fail("PropertyObject::SetValue error");
                Logger.LogHandler.Error(ex.Message);
            }
        }

        public object GetValue()
        {
            try
            {
                if (m_prop != null)
                {
                    return m_prop.GetValue(m_obj, null);
                }
            }
            catch (System.Exception ex)
            {
                System.Diagnostics.Debug.Fail("PropertyObject::GetValue error");
                Logger.LogHandler.Error(ex.Message);
            }
            return null;
        }

        public void SetValueByString(String strValue)
        {
            try
            {
                if (m_bIsOk)
                {
                    Object objValue = null;

                    switch (m_eDataType)
                    {
                        case DataType.IOLib_Base_EquipmentStatus:
                            {
                                objValue = (Boolean)true;
                                Boolean bValue = false;
                                if (Boolean.TryParse(strValue, out bValue))
                                {
                                    objValue = bValue ? EquipmentStatus.OPENED : EquipmentStatus.CLOSED;
                                }
                                else
                                {
                                    objValue = EquipmentStatus.UNKNOWN;
                                }
                            }
                            break;
                        case DataType.System_Single:
                            {
                                Single fValue = 0.0f;
                                if (Single.TryParse(strValue, out fValue))
                                {
                                    objValue = fValue;
                                }
                            }
                            break;
                        case DataType.System_Boolean:
                            {
                                Boolean bValue = false;
                                if (Boolean.TryParse(strValue, out bValue))
                                {
                                    objValue = bValue;
                                }
                            }
                            break;
                    }

                    if (objValue != null)
                    {
                        m_prop.SetValue(m_obj, objValue, null);
                    }
                }
            }
            catch (Exception ex)
            {
                System.Diagnostics.Debug.Fail("PropertyObject::SetValueByString error");
                Logger.LogHandler.Error(ex.Message);
            }
        }

        private DataType GetDataType(String strDataType)
        {
            DataType eDataType = DataType.Undefined;
            if (strDataType == "IOLib.Base.EquipmentStatus")
            {
                eDataType = DataType.IOLib_Base_EquipmentStatus;
            }
            else if (strDataType == "System.Single")
            {
                eDataType = DataType.System_Single;
            }
            else if (strDataType == "System.Boolean")
            {
                eDataType = DataType.System_Boolean;
            }
            else
            {
                eDataType = DataType.Undefined;
                System.Diagnostics.Debug.Fail("Not support data type:" + m_prop.PropertyType.FullName);
            }
            return eDataType;
        }
    }
}
